composinator/convert.go

```diff
@@ -0,0 +1,517 @@
+)
+
+func convert(w http.ResponseWriter, client *v3.RancherClient, input convertOptions) {
+	services, volumeTemplates, certMap, serviceMap, containerMap, secretMap, err := prepareExportedData(client, input.StackID)
+	if err != nil {
+		http.Error(w, "can't obtain exported data", http.StatusInternalServerError)
+	}
+	dockerCompose, rancherCompose, compose, err := createComposeData(services, volumeTemplates, certMap, serviceMap, containerMap, secretMap, input.Format)
+	if err != nil {
+		http.Error(w, "can't create compose file", http.StatusInternalServerError)
+	}
+	result := map[string]string{}
+	if input.Format == "combined" {
+		result["compose"] = compose
+	} else {
+		result["dockerCompose"] = dockerCompose
+		result["rancherCompose"] = rancherCompose
+	}
+	w.WriteHeader(http.StatusOK)
+	data, err := json.Marshal(result)
+	if err != nil {
+		http.Error(w, "can't marshall result", http.StatusInternalServerError)
+	}
+	fmt.Fprint(w, data)
+}
+
+func prepareExportedData(client *v3.RancherClient, stackID string) ([]*v3.Service, map[string]v3.VolumeTemplate, map[string]v3.Certificate, map[string]v3.Service, map[string]v3.Instance, map[string]v3.Secret, error) {
+	volumeTemplates, err := client.VolumeTemplate.List(&v3.ListOpts{
+		Filters: map[string]interface{}{
+			"stackId":      stackID,
+			"removed_null": "true",
+			"limit":        "-1",
+		},
+	})
+	if err != nil {
+		return nil, nil, nil, nil, nil, nil, errors.Wrap(err, "can't list volumeTemplates")
+	}
+	stack, err := client.Stack.ById(stackID)
+	if err != nil {
+		return nil, nil, nil, nil, nil, nil, errors.Wrap(err, "can't get stack")
+	}
+	servicesToExport := []*v3.Service{}
+	// lbcertMap is used for certificate reference by lbConfig
+	lbcertMap := map[string]v3.Certificate{}
+	// serviceMap is used for reference by portRule
+	serviceMap := map[string]v3.Service{}
+	containerMap := map[string]v3.Instance{}
+	secretMap := map[string]v3.Secret{}
+	for _, id := range stack.ServiceIds {
+		service, err := client.Service.ById(id)
+		if err != nil {
+			return nil, nil, nil, nil, nil, nil, errors.Wrap(err, "can't get service")
+		}
+		if service.LbConfig != nil {
+			// prepare certMap
+			if service.LbConfig.DefaultCertificateId != "" {
+				cert, err := client.Certificate.ById(service.LbConfig.DefaultCertificateId)
+				if err != nil {
+					return nil, nil, nil, nil, nil, nil, errors.Wrap(err, "can't get lbconfig")
+				}
+				lbcertMap[cert.Id] = *cert
+			}
+			if len(service.LbConfig.CertificateIds) > 0 {
+				for _, certID := range service.LbConfig.CertificateIds {
+					cert, err := client.Certificate.ById(certID)
+					if err != nil {
+						return nil, nil, nil, nil, nil, nil, errors.Wrap(err, "can't get lbconfig")
+					}
+					lbcertMap[cert.Id] = *cert
+				}
+			}
+			// prepare serviceMap and containerMap
+			for _, portRule := range service.LbConfig.PortRules {
+				if portRule.ServiceId != "" {
+					service, err := client.Service.ById(portRule.ServiceId)
```

Comment:
For these you should check first if you already have the serviceId/containerId in one of the lists you already downloaded.  If not then do ById.